### ISSUE TITLE
Dynamic texture selection is impossible for sample() calls

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#2482]
In MSL code, it's pretty common to write code like this:

```
    texture2d<float> texture;
    sampler sampler;

    if (nonuniform)
    {
        texture = textureA;
        sampler = samplerA;
    } else {
        texture = textureB;
        sampler = samplerB;
    }

    return texture.sample(sampler, textureCoordinate);
```

This is legal because sampling only requires uniform control flow - not a uniform texture.

If you wanted to write this kind of code in WGSL, it seems to be impossible.

1. You can't make a local variable of texture type, like this example does, because https://gpuweb.github.io/gpuweb/wgsl/#storage-class says that variables in the `handle` storage class can only exist at module scope.
2. You can't make a pointer to a texture type, because https://gpuweb.github.io/gpuweb/wgsl/#address-of says "It is a shader-creation error if SC is the handle storage class."
3. You can't move the `sample()` call up into the conditional because sampling requires uniform control flow.

This is a request coming from one of our internal teams - it would be unfortunate if WGSL was unable to support this kind of use.

**Comment by alan-baker** [comment]
SPIR-V does not support selections between samplers or images. Instead of the given code, authors would be expected to have multiple sample calls that are dynamically uniform. That is at odds with the desire of the uniformity analysis so we may be stuck on not allowing this type of behaviour.

In fact, Vulkan requires that all descriptor accesses (e.g. buffers, images, samplers) are dynamically uniform or annotated as non-uniform. The non-uniform annotation entails an extra feature which I don't believe is currently required to use WebGPU.
>If an instruction loads from or stores to a resource (including atomics and image instructions) and the resource descriptor being accessed is not dynamically uniform, then the corresponding non-uniform indexing feature must be enabled and the capability must be declared.

**Comment by kvark** [comment]
This code is semantically equivalent to :
```cpp
    if (nonuniform) {
        return textureA.sample(samplerA, textureCoordinate);
    } else {
        return textureB.sample(samplerB, textureCoordinate);
    }
```
Based on the conservative rules we have it's indeed invalid. But I think @litherum has a interesting point that this code should perhaps be fully safe and portable?

I'll try to elaborate. Sampling in non-uniform control flow is bad because derivatives may not exist for texture coordinates. This may be the case if other invocations are early terminated, for example. But here, *both* branches use `textureCoordinate`, so we can statically prove that `textureCoordinate` is computed on all invocations, and therefore the derivatives are portable.

@alan-baker are you sure the quoted segments applies to single descriptors? It sounds to be about descriptor indexing specifically, where the index needs to be dynamically uniform. Otherwise, even code like this doesn't fall under the requirements posted:
```
if (non_dynamically_uniform) {
  let foo = my_buffer.data;
}
```
And this code is valid in WGSL today.

**Comment by alan-baker** [comment]
The quote applies generally, but since you can't do an OpSelect or OpPhi on images and samplers the only way to trigger it is with a resource array and descriptor indexing.

The code you quote at the end is perfectly fine because the access is dynamically uniform. Every invocation that enters the if is accessing the same resource, but not every invocation enters the if.

For the analysis of `textureCoordinate` to work, I think you have to assume that the hardware works a certain way that can't be guaranteed. That is, you seem to assume the quad invocations will all process both texture operations. That might be true  for some hardware, but I don't think it would work everywhere.

**Comment by litherum** [comment]
I think your reasoning is invalid, @kvark. Because GPUs are allowed to have a simdgroup (running in lock-step) execute both sides of a branch, the code you wrote can be executed as two serialized sample calls, rather than one sample call where half the threads are on one side of the branch and the other half are on the other side.

If the threads are running in lock-step, and half the threads are computing a derivative, they can grab texture coordinate values from the threads which are inactive (but still here along for the ride), which will produce bogus results.

So, hoisting and duplicating the sample call up into the conditional is a fundamentally different, and incorrect, program.

**Comment by kvark** [comment]
Ok, well, in this case the internal teams can do this instead:
```rust
let ddx = dpdx(textureCoordinate);
let ddy = dpdy(textureCoordinate);
if (nonuniform) {
        return textureA.sampleGrad(samplerA, textureCoordinate, ddx, ddy);
    } else {
        return textureB.sampleGrad(samplerB, textureCoordinate, ddx, ddy);
    }
```

It's not as optimal, but at least portable (e.g. to SPIR-V), and supported today on WGSL.

**Comment by litherum** [comment]
Yeah, I think the code in https://github.com/gpuweb/gpuweb/issues/2482#issuecomment-1007567396 is a correct workaround, but it's pretty yucky. If SPIR-V really just doesn't allow this feature (which is surprising to hear, given that this pattern is pretty common in MSL), then it seems the path forward would be to make this kind of dynamic selection into an optional WebGPU feature (probably for after MVP).

@alan-baker

**Comment by kvark** [comment]
I think Alan meant `shaderSampledImageArrayNonUniformIndexing` in either [VkPhysicalDeviceVulkan12Features](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPhysicalDeviceVulkan12Features.html) or [VkPhysicalDeviceDescriptorIndexingFeaturesEXT](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkPhysicalDeviceDescriptorIndexingFeaturesEXT.html). It's a bit strange that this feature would apply to non-arrays as well.

GpuDB lists support for [95.6% reports](https://vulkan.gpuinfo.org/listdevicescoverage.php?core=1.2&feature=shaderSampledImageArrayNonUniformIndexing&platform=all), which is pretty strong!

**Comment by alan-baker** [comment]
It's more complicated perhaps than I made it sound. SPIR-V simply doesn't allow dynamic selection (via OpSelect or OpPhi) of images or samplers. So compilers would have to try to use descriptor indexing (the feature I mentioned) instead. That means WGSL compilers to Vulkan would have to change the bindings written in the shader to remap to a resource array instead which wouldn't be a fun translation.

**Comment by magcius** [comment]
There's a third option:

```wgsl
    vec4<f32> sampleA = textureA.sample(samplerA, textureCoordinate);
    vec4<f32> sampleB = textureB.sample(samplerB, textureCoordinate);

    if (nonuniform) {
        return sampleA;
    } else {
        return sampleB;
    }
```

Obviously not as efficient as other approaches.

Not all GPUs can easily support the code @litherum suggests naturally; there are some mobile GPUs that submit their texture to the texture unit at the subgroup level, and needs to be uniform in that packet. This is why `NonUniformResourceIndex` is an optional feature on a lot of devices now, and why it requires an explicit function call in HLSL; the shader compiler might have to do some extra work to work around it.

A similar restriction is implied today in WGSL through other means, but we might want to make it explicit, with a possible extension in the future to relax it.

**Comment by dneto0** [comment]
> That means WGSL compilers to Vulkan would have to change the bindings written in the shader to remap to a resource array instead which wouldn't be a fun translation.

That would be a horrible terrible no-good option.

**Comment by dneto0** [comment]
> GpuDB lists support for [95.6% reports](https://vulkan.gpuinfo.org/listdevicescoverage.php?core=1.2&feature=shaderSampledImageArrayNonUniformIndexing&platform=all), which is pretty strong! Edit: some of these reports are for older drivers (e.g. AMD rx570 for 2018), while the newer driver supports the feature.

I was puzzled by this metric.   My prior expectation was that it would be common on desktop and rare on Android.

That 95.6% number seems really high.  I think for the denominator it's using "all device reports for Vulkan 1.2". But Vulkan 1.2 is rare on Android.
The feature was only available as an extension in Vulkan 1.1., and looking through that lens ([link](https://vulkan.gpuinfo.org/listfeaturesextensions.php)) the gpu database reports 41.9% coverage. 

I'm not asserting this has great bearing on the outcome of this issue.  I just want to caution that use of that gpu info database needs some amount of care.  (And yeah, this ties into #1069: we should be more deliberate about picking a set of representative devices on the Android side.)

**Comment by kdashg** [comment]
Post-v1 based on https://github.com/gpuweb/gpuweb/issues/2482#issuecomment-1007571743

**Comment by stefnotch** [comment]
It occurred to me that there's another legal transformation. One can mechanically take the rest of the code in the function, and extract it into a separate function. That actually scales to an arbitrary number of textures, it just introduces one "continuation" function every time.
```wgsl
fn main() {
    texture2d<float> texture;
    sampler sampler;

    if (nonuniform) {
       return sample_texture_and_continue(textureA, samplerA);
    } else {
       return sample_texture_and_continue(textureB, samplerB);
    }
}

fn sample_texture_and_continue(texture2d<float> texture, sampler sampler) {
    return texture.sample(sampler, textureCoordinate);
}
```


That aside, why does WGSL currently not allow assigning a texture to a `let`, or returning a texture from a function? Would it be technically feasible to eventually relax some of those restrictions?
```wgsl
let a: texture_2d<f32> = my_texture;
```

```wgsl
fn foo() -> texture_2d<f32> { ... }
```

**Comment by magcius** [comment]
See also https://github.com/gpuweb/gpuweb/issues/5085
